providers/oauth2: Add provider federation between OAuth2 Providers (#12083)
goauthentik/authentik: website/docs/add-secure-apps/providers/oauth2/client_credentials.md

### JWT-authentication

#### Externally issued JWTs <span class="badge badge--version">authentik 2022.4+</span>

You can authenticate and get a token using an existing JWT. For readability we will refer to the JWT issued by the external issuer/platform as input JWT, and the resulting JWT from authentik as the output JWT.

> To configure this, define a JWKS URL/raw JWKS data in OAuth Sources. If a JWKS URL is specified, authentik will fetch the data and store it in the source, and then select the source in the OAuth2 Provider that will be authenticated against.

Reviewer: data source?  "...define a JWKS URL/raw JWKS data in OAuth Sources." is an incomplete sentence, needs the noun.

> With this configure, any JWT issued by the configured sources' certificates can be used to authenticate:

Reviewer: ```suggestion
With this configuration, any JWT issued by the configured sources' certificates can be used to authenticate:
```

```
POST /application/o/token/ HTTP/1.1
Host: authentik.company
Content-Type: application/x-www-form-urlencoded

grant_type=client_credentials&
client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer&
client_assertion=$inputJWT&
client_id=application_client_id
```

> Alternatively, you can set the `client_secret` parameter to the `$inputJWT`, for applications which can set the password from a file but not other parameters.

Reviewer: ```suggestion
Alternatively, you can set the `client_secret` parameter to `$inputJWT`, for applications that can set the password from a file but not other parameters.
```

Input JWTs are checked to be signed by any of the selected _Federated OIDC Sources_, and their `exp` attribute must not be now or in the past.

To dynamically limit access based on the claims of the tokens, you can use _[Expression policies](../../../customize/policies/expression.mdx)_:

```python
return request.context["oauth_jwt"]["iss"] == "https://my.issuer"
```

#### authentik-issued JWTs <span class="badge badge--version">authentik 2024.12+</span>

> To allow federation between providers, select the provider of the application you want to federate to in the provider settings of the application whose token will be used for authentication.

Reviewer: ```suggestion
To allow federation between providers, modify the provider settings of the application (whose token will be used for authentication) to select the provider of the application to which you want to federate.
```

Reviewer: Phew, that's a hard one. There might be even better wording but this will do for now I think.
